rvgo/fast/radix.go

```diff
@@ -0,0 +1,444 @@
+type L8 = *LargeRadixNode[L9]
+type L9 = *LargeRadixNode[L10]
+type L10 = *LargeRadixNode[L11]
+type L11 = *Memory
+
+// InvalidateNode invalidates the hash cache along the path to the specified address.
+// It marks the necessary child hashes as invalid, forcing them to be recomputed when needed.
+func (n *SmallRadixNode[C]) InvalidateNode(addr uint64) {
+	childIdx := addressToRadixPath(addr, n.Depth, 4) // Get the 4-bit child index at the current depth.
+
+	branchIdx := (childIdx + 1<<4) / 2 // Compute the index for the hash tree traversal.
+
+	// Traverse up the hash tree, invalidating hashes along the way.
+	for index := branchIdx; index > 0; index >>= 1 {
+		hashBit := index & 15          // Get the relevant bit position (0-15).
+		n.ChildExists |= 1 << hashBit  // Mark the child as existing.
+		n.HashValid &= ^(1 << hashBit) // Invalidate the hash at this position.
+	}
+}
+
+func (n *LargeRadixNode[C]) InvalidateNode(addr uint64) {
+	childIdx := addressToRadixPath(addr, n.Depth, 8)
+
+	branchIdx := (childIdx + 1<<8) / 2
+
+	for index := branchIdx; index > 0; index >>= 1 {
+		hashIndex := index >> 6
+		hashBit := index & 63
+		n.ChildExists[hashIndex] |= 1 << hashBit
+		n.HashValid[hashIndex] &= ^(1 << hashBit)
+	}
+}
+
+func (m *Memory) InvalidateNode(addr uint64) {
+	if p, ok := m.pageLookup(addr >> PageAddrSize); ok {
+		p.Invalidate(addr & PageAddrMask)
+	}
+}
+
+// GenerateProof generates the Merkle proof for the given address.
+// It collects the necessary sibling hashes along the path to reconstruct the Merkle proof.
+func (n *SmallRadixNode[C]) GenerateProof(addr uint64) [][32]byte {
+	var proofs [][32]byte
+	path := addressToRadixPath(addr, n.Depth, 4)
+
+	if n.Children[path] == nil {
+		// When no child exists at this path, the rest of the proofs are zero hashes.
+		proofs = zeroHashRange(0, 60-n.Depth-4)
+	} else {
+		// Recursively generate proofs from the child node.
+		proofs = (*n.Children[path]).GenerateProof(addr)
+	}
+
+	// Collect sibling hashes along the path for the proof.
+	for idx := path + 1<<4; idx > 1; idx >>= 1 {
+		sibling := idx ^ 1 // Get the sibling index.
+		proofs = append(proofs, n.MerkleizeNode(addr>>(64-n.Depth), sibling))
+	}
+
+	return proofs
+}
+
+func (n *LargeRadixNode[C]) GenerateProof(addr uint64) [][32]byte {
+	var proofs [][32]byte
+	path := addressToRadixPath(addr, n.Depth, 8)
+
+	if n.Children[path] == nil {
+		proofs = zeroHashRange(0, 60-n.Depth-8)
+	} else {
+		proofs = (*n.Children[path]).GenerateProof(addr)
+	}
+
+	for idx := path + 1<<8; idx > 1; idx >>= 1 {
+		sibling := idx ^ 1
+		proofs = append(proofs, n.MerkleizeNode(addr>>(64-n.Depth), sibling))
+	}
+	return proofs
+}
+
+func (m *Memory) GenerateProof(addr uint64) [][32]byte {
+	pageIndex := addr >> PageAddrSize
+
+	if p, ok := m.pages[pageIndex]; ok {
+		return p.GenerateProof(addr) // Generate proof from the page.
+	} else {
+		return zeroHashRange(0, 8) // Return zero hashes if the page does not exist.
```

Comment:
It's because the number of proof for a page is 8, where 0 is the leaf page data and 7 is the page's merkle root. I've also added this as a comment.